### ISSUE TITLE
AssertionError

### ISSUE THREAD

**Issue by dmitrikudrenko** [smooch/smooch-android#13]
I use last version of smooch libraries (2.1.8). After Smooch.init I get that error

```
Fatal Exception: java.lang.AssertionError
       at com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter.(Unknown Source)
       at com.google.gson.internal.bind.TypeAdapters$26.create(Unknown Source)
       at com.google.gson.Gson.doubleAdapter(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.excludeField(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.excludeField(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.excludeField(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.excludeField(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.excludeField(Unknown Source)
       at com.google.gson.Gson.doubleAdapter(Unknown Source)
       at com.google.gson.internal.bind.CollectionTypeAdapterFactory.create(Unknown Source)
       at com.google.gson.Gson.doubleAdapter(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.excludeField(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.excludeField(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.excludeField(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.excludeField(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.excludeField(Unknown Source)
       at com.google.gson.Gson.doubleAdapter(Unknown Source)
       at com.google.gson.Gson.doubleAdapter(Unknown Source)
       at com.google.gson.Gson.doubleAdapter(Unknown Source)
       at com.google.gson.Gson.floatAdapter(Unknown Source)
       at com.google.gson.Gson.floatAdapter(Unknown Source)
       at io.smooch.core.facade.impl.h.a(Unknown Source)
       at io.smooch.core.facade.c.a(Unknown Source)
       at io.smooch.core.service.SmoochService.k(Unknown Source)
       at io.smooch.core.i.run(Unknown Source)
       at io.smooch.core.e.a(Unknown Source)
       at io.smooch.core.e.i(Unknown Source)
       at io.smooch.core.User.getCurrentUser(Unknown Source)
       at io.smooch.core.User.getCurrentUser(Unknown Source)
       at io.smooch.core.p.run(Unknown Source)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:148)
       at android.app.ActivityThread.main(ActivityThread.java:5417)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

**Comment by jugarrit** [comment]
Can you please add some steps to reproduce the problem then re-open this issue?

Thanks :)

**Comment by nizamsp** [comment]
In my production version, 89 crashes happened with the same error.
All these crashes on background (app was not in focus) and all are on non-rooted phones.

Traceback below:
```
Fatal Exception: java.lang.AssertionError
       at com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter.(Unknown Source)
       at com.google.gson.internal.bind.TypeAdapters$26.create(Unknown Source)
       at com.google.gson.Gson.getAdapter(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getFieldAdapter(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.access$100(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.createBoundField(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(Unknown Source)
       at com.google.gson.Gson.getAdapter(Unknown Source)
       at com.google.gson.internal.bind.CollectionTypeAdapterFactory.create(Unknown Source)
       at com.google.gson.Gson.getAdapter(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getFieldAdapter(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.access$100(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.createBoundField(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(Unknown Source)
       at com.google.gson.Gson.getAdapter(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getFieldAdapter(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.access$100(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.createBoundField(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(Unknown Source)
       at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(Unknown Source)
       at com.google.gson.Gson.getAdapter(Unknown Source)
       at com.google.gson.Gson.fromJson(Unknown Source)
       at com.google.gson.Gson.fromJson(Unknown Source)
       at com.google.gson.Gson.fromJson(Unknown Source)
       at com.google.gson.Gson.fromJson(Unknown Source)
       at io.smooch.core.facade.impl.a.a(Unknown Source)
       at io.smooch.core.facade.i.a(Unknown Source)
       at io.smooch.core.facade.impl.f.run(Unknown Source)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:145)
       at android.app.ActivityThread.main(ActivityThread.java:5832)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)
```

**Comment by jugarrit** [comment]
On which device?

**Comment by nizamsp** [comment]
Its happening mainly in Samsung Phones.

sm_g900h - 35%
sm_n900 - 27%
sm_a500g - 24%

**Comment by dmitrikudrenko** [comment]
@nizamsp check your proguard. It solved my problem.

**Comment by nizamsp** [comment]
Thank you @dmitrikudrenko. My Proguard is like the following. Am I missing something here.
```
-optimizationpasses 5
-dontusemixedcaseclassnames
-dontskipnonpubliclibraryclasses
-dontpreverify
-verbose
-optimizations !code/simplification/arithmetic,!field/*,!class/merging/*

-keep public class * extends android.app.Activity
-keep public class * extends android.app.Application
-keep public class * extends android.app.Service
-keep public class * extends android.content.BroadcastReceiver
-keep public class * extends android.content.ContentProvider
-keep public class * extends android.app.backup.BackupAgentHelper
-keep public class * extends android.preference.Preference
-keep public class com.android.vending.licensing.ILicensingService
-keep public class * implements java.io.Serializable
-keep public class * extends android.support.v4.app.Fragment
-keep public class * extends android.support.v4.app.ListFragment



-keep public class org.spongycastle.** {
  <fields>;
  <methods>;
}

-keep public class org.apache.** {
  <fields>;
  <methods>;
}

-ignorewarnings

-keepclasseswithmembernames class * {
    native <methods>;
}

-keepclasseswithmembers class * {
    public <init>(android.content.Context, android.util.AttributeSet);
}

-keepclasseswithmembers class * {
    public <init>(android.content.Context, android.util.AttributeSet, int);
}

-keepclassmembers class * extends android.app.Activity {
   public void *(android.view.View);
}


-keepclassmembers enum * {
    public static **[] values();
    public static ** valueOf(java.lang.String);
}

-keep class * implements android.os.Parcelable {
  public static final android.os.Parcelable$Creator *;
}

-keepclassmembers class * implements java.io.Serializable {
    static final long serialVersionUID;
    private static final java.io.ObjectStreamField[] serialPersistentFields;
    private void writeObject(java.io.ObjectOutputStream);
    private void readObject(java.io.ObjectInputStream);
    java.lang.Object writeReplace();
    java.lang.Object readResolve();
}

-keep public class * extends android.app.Fragment

-keep class de.greenrobot.** {*;}

-keepclassmembers class ** {
    public void onEvent*(**);
}

-keep class com.google.gson.** { *; }
-keep class com.google.inject.** { *; }
-keep class org.apache.http.** { *; }
-keep class org.apache.james.mime4j.** { *; }
-keep class javax.inject.** { *; }
-keep class retrofit.** { *; }
-keep class com.whereismytrain.** { *; }
-keep class com.uberapi.** { *; }
-keep class com.ubermodel.** { *; }
-keep class com.wimtwear.** { *; }
-dontwarn rx.**
-keep class com.example.testobfuscation.** { *; }
-keepattributes Signature
-keep class sun.misc.Unsafe { *; }
-keepclassmembers enum * { *; }

-keepattributes *Annotation*,SourceFile,LineNumberTable
```

**Comment by nizamsp** [comment]
Why was this issue closed when its not resolved?

**Comment by skolarianer** [comment]
The following does the trick for me...
Found [here](http://stackoverflow.com/a/37125936/6619198).

```
-dontwarn okio.**
-keep class com.google.gson.** { *; }
-keepclassmembers enum * { *; }
-keepattributes Signature
-keepattributes *Annotation*
```

But I agree to @nizamsp. This should be done in the library.

**Comment by nizamsp** [comment]
Yes. I am missing the following one in my progaurd.
-keepclassmembers enum * { *; }

**Comment by jugarrit** [comment]
Just pushed a proguard fix in 4.0.5
Hope that helps :)

**Comment by skolarianer** [comment]
Which rules have you implemented? It keeps crashing when I remove the rules mentioned above from my proguard config.

**Comment by jugarrit** [comment]
```
-dontwarn okio.**
-keep class com.google.gson.** { *; }
-keepattributes Signature
-keepattributes *Annotation*
-keepclassmembers enum io.smooch.core.model.** { *; }
```

**Comment by skolarianer** [comment]
Ok, strange. These settings are not fetched. At least for me.

**Comment by skolarianer** [comment]
I can't find any proguard configuration in your 'aar'.

![noproguard](https://cloud.githubusercontent.com/assets/6286711/18289163/8c90612e-747e-11e6-9a6a-ccd4157d667a.png)

**Comment by jugarrit** [comment]
@skolarianer Sorry about that...

I was under the impression I didn't need to provide `consumerProguardFiles` for Smooch to function correctly if I had these settings in our proguard file, but I now realize my error.

I will be sure to fix this in the coming days.

**Comment by skolarianer** [comment]
`-keepclassmembers enum io.smooch.core.model.** { *; }` seems to be not enough. I still need to add `-keepclassmembers enum * { *; }`. Otherwise the AssertionError will be thrown again.

I tried to add `-keepclassmembers enum com.google.gson.** { *; }` only, but that didn't help.

**Comment by jugarrit** [comment]
Thanks for all the help!

**Comment by wmora** [comment]
proguard rules are included with the SDK since version [5.0.0](https://github.com/smooch/smooch-android/releases/tag/5.0.0)
